src/qutip_tensorflow/core/data/tftensor.py

```diff
@@ -79,13 +80,16 @@ def to_array(self):
         return self._tf.numpy()
 
     def conj(self):
-        return TfTensor(tf.math.conj(self._tf))
+        from .adjoint import conj_tftensor
```

Comment:
Importing here is not nice looking.
It would be nicer to have just `adjoint.py` import `tftensor.py` and not the other way around.
I would prefer to have the method and function calling tensorflow directly or `adjoint_tftensor` using the method.

Reply:
I added it here due to a circular importing error I was having but I agree it does not look good there. I will make the function to call tensorflow directly and add proper tests for these methods. 

I know I am repeating myself here but having two ways of doing the same thing (methods and specialisations) does not look like a good idea as it is a source of possible conflict (as the python bible says "There should be one-- and preferably only one --obvious way to do it.").

Reply:
I don't think Eric's suggesting that you duplicate the code, more that you put the specialisation code in the method instead.  That said, I don't really agree that it's a problem to have the import here.  The Cython datalayer types don't have to do it because Cython handles the linking at compile time, but they do effectively the same thing.

The zen of Python (it's not a bible by any stretch) is more a rough set of guidelines, and having a method and a similar function is _very_ common.  See: all of Numpy, `list.extend` and `list.__add__`, and many others.

Reply:
Here, code duplication does not bother me since it's only one line calling a tensorflow function. But having the function use the method, or having both the function and method in the same file seems cleaner to me than having a circular import.